sdks/java/core/src/test/java/org/apache/beam/sdk/util/ApiSurfaceVerificationTest.java

```diff
@@ -0,0 +1,244 @@
+
+    /* <helper_lambdas> */
+
+    final Function<Class<?>, List<Class<?>>> toExposure =
+        new Function<Class<?>, List<Class<?>>>() {
+
+          @Override
+          public List<Class<?>> apply(@Nonnull final Class<?> aClass) {
+            return checkedApiSurface.getAnyExposurePath(aClass);
+          }
+        };
+
+    final Maps.EntryTransformer<Class<?>, List<Class<?>>, String> toMessage =
+        new Maps.EntryTransformer<Class<?>, List<Class<?>>, String>() {
+
+          @Override
+          public String transformEntry(@Nonnull final Class<?> aClass,
+                                       @Nonnull final List<Class<?>> exposure) {
+            return aClass
+                + " exposed via:\n\t\t"
+                + Joiner.on("\n\t\t").join(exposure);
+          }
+        };
+
+    final Predicate<Class<?>> disallowed = new Predicate<Class<?>>() {
+
+      @Override
+      public boolean apply(@Nonnull final Class<?> aClass) {
+        return !classIsAllowed(aClass, allowedPackages);
+      }
+    };
+
+    /* </helper_lambdas> */
+
+    final FluentIterable<Class<?>> disallowedClasses =
+        FluentIterable
+            .from(checkedApiSurface.getExposedClasses())
+            .filter(disallowed);
+
+    final ImmutableMap<Class<?>, List<Class<?>>> exposures =
+        Maps.toMap(disallowedClasses, toExposure);
+
+    final ImmutableList<String> messages =
+        FluentIterable
+            .from(Maps.transformEntries(exposures, toMessage).values())
+            .toSortedList(Ordering.<String>natural());
+
+    if (!messages.isEmpty()) {
+      fail("The following disallowed classes appear in the public API surface:\n\t"
+               + Joiner.on("\n\t").join(messages));
+    }
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private boolean classIsAllowed(final Class<?> clazz,
+                                 final Set<Matcher<Class<?>>>  allowedPackages) {
+    // Safe cast inexpressible in Java without rawtypes
+    return anyOf((Iterable) allowedPackages).matches(clazz);
+  }
+
+  private void assertApiSurface(final ApiSurface checkedApiSurface,
+                                final Set<Matcher<Class<?>>> allowedPackages)
+      throws Exception {
+
+    assertDisallowed(checkedApiSurface, allowedPackages);
+    assertAbandoned(checkedApiSurface, allowedPackages);
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  protected Set<String> prunePatterns() {
+    return Sets.newHashSet("org[.]apache[.]beam[.].*Test.*",
+                           "org[.]apache[.]beam[.].*IT",
+                           "java[.]lang.*");
+  }
+
+  protected ApiSurface apiSurface() throws IOException {
+    final String thisPackage = getClass().getPackage().getName();
+    LOG.info("Verifying the surface of API package: {}", thisPackage);
+    final ApiSurface apiSurface = ApiSurface.ofPackage(thisPackage);
+    return prune(apiSurface, prunePatterns());
+  }
+
+  @Test
+  public void testApiSurface() throws Exception {
```

Comment:
This might actually work well as a `Matcher<ApiSurface>` along the lines of `assertThat(sdkApiSurface, containsOnlyClassesMatching(allowedClassesMatcher()))`